packages/node/src/message-handling/handle-protocol-message.ts

```diff
@@ -36,138 +37,213 @@ export async function handleReceivedProtocolMessage(
-      | TakeActionParams
-      | UpdateParams;
-
-    await requestHandler.getShardedQueue(multisigAddress).add(async () => {
-      stateChannelsMap = await instructionExecutor.runProtocolWithMessage(
-        data,
-        new Map<string, StateChannel>(
-          Object.entries(await store.getAllChannels())
+  const outgoingEventData = getOutgoingEventDataFromProtocol(
+    protocol,
+    params,
+    publicIdentifier,
+    postProtocolStateChannelsMap
+  );
+
+  if (outgoingEventData) {
+    await emitOutgoingNodeMessage(router, outgoingEventData);
+  }
+}
+
+function emitOutgoingNodeMessage(router: RpcRouter, msg: object) {
+  return router.emit(msg["type"], msg, "outgoing");
+}
+
+function getOutgoingEventDataFromProtocol(
+  protocol: string,
+  params: ProtocolParameters,
+  publicIdentifier: string,
+  stateChannelsMap: Map<string, StateChannel>
+) {
+  const baseEvent = { from: publicIdentifier };
+
+  switch (protocol) {
+    case Protocol.Install:
+      // TODO: Have to take an InstallParams object and somehow compute the
+      //       appInstanceIdentityHash from it and then emit an event with
+      //       that value inside of it.
+      return;
+    case Protocol.Uninstall:
+      return {
+        ...baseEvent,
+        type: NODE_EVENTS.UNINSTALL,
+        data: getUninstallEventData(params as UninstallParams)
+      };
+    case Protocol.Setup:
+      return {
+        ...baseEvent,
+        type: NODE_EVENTS.CREATE_CHANNEL,
+        data: getSetupEventData(
+          params as SetupParams,
+          stateChannelsMap.get((params as SetupParams).multisigAddress)!
+            .multisigOwners
         )
+      };
+    case Protocol.Withdraw:
+      return {
+        ...baseEvent,
+        type: NODE_EVENTS.WITHDRAWAL_CONFIRMED,
+        data: getWithdrawEventData(params as WithdrawParams)
+      };
+    case Protocol.TakeAction:
+    case Protocol.Update:
+      return {
+        ...baseEvent,
+        type: NODE_EVENTS.UPDATE_STATE,
+        data: getStateUpdateEventData(
+          params as UpdateParams,
+          stateChannelsMap
+            .get((params as TakeActionParams | UpdateParams).multisigAddress)!
+            .getAppInstance(
+              (params as TakeActionParams | UpdateParams).appIdentityHash
+            )!.state
+        )
+      };
+    case Protocol.InstallVirtualApp:
+      // TODO: Have to take an InstallParams object and somehow compute the
+      //       appInstanceIdentityHash from it and then emit an event with
+      //       that value inside of it.
+      return;
+    case Protocol.UninstallVirtualApp:
+      return {
+        ...baseEvent,
+        type: NODE_EVENTS.UNINSTALL_VIRTUAL,
+        data: getUninstallVirtualAppEventData(
+          params as UninstallVirtualAppParams
+        )
+      };
+    default:
+      throw new Error(
+        `handleReceivedProtocolMessage received invalid protocol message: ${protocol}`
       );
+  }
+}
+
+const getStateUpdateEventData = (
```

Comment:
Code minimization; no margin for adding mutation or anything, just returns precisely one thing in a very simple statement

Reply:
An implementation with the function syntax wouldn't add mutation either. I'm just advocating for that to keep the style consistent, not because I think what this is doing is not simple.